tools/roadmap/validate_dag.py

```
@@ -0,0 +1,119 @@
+        return 3
+
+    # Optional: detect cycles in hard-edge graph (DAG requirement)
+    adj: Dict[str, List[str]] = {}
+    nodes_set = {n for (n, _) in nodes}
+    for n in nodes_set:
+        adj.setdefault(n, [])
+    for s, d in edges_hard:
+        adj.setdefault(s, []).append(d)
+
+    WHITE, GRAY, BLACK = 0, 1, 2
+    color: Dict[str, int] = {n: WHITE for n in nodes_set}
```

Ruff (0.14.3)

Unnecessary dict comprehension for iterable; use `dict.fromkeys` instead

Replace with `dict.fromkeys(iterable)`)

(C420)